### DISCUSSION TITLE
How can I extend session expiry?

### DISCUSSION THREAD

**Discussion by byte-my-briefs** [clerk/javascript#3021] [Support request - DO NOT OPEN HERE]
# Brief Question

I would like to extend the session expiry (`"exp"`). Currently, it seems to be set to 60 seconds. For reasons explained at more length in the next section, I would like to extend it rather significantly. 

The session docs claim that the `"exp"` field is ["[d]etermined using the Token lifetime JWT template setting in the Clerk Dashboard."](https://clerk.com/docs/backend-requests/resources/session-tokens). However, I modified this up to 604,800 (1 week) to test it yesterday and it has had no effect. 

**Does anyone know how I can address this?** 

# Detailed Summary of Issue

_NOTE: The above is a workaround fix for the issue I am experiencing below. The below is more complex, and I welcome any advice that helps to address the issue more directly._

I have a (1) Webapp, and (2) a Chrome extension (primarily run as a "New Tab"), both coded in Sveltekit. I would like to share session state between the two. This works, so long as the webapp is open in a separate tab. However, after the webapp is closed, the session within the extension expires with 60 seconds. Though the frontend session is still recognize (and the `<SignedIn />` component therefore still functions properly), the backend rejects all requests, claiming that the session token ahs expired. 

I have tried adapting the [@clerk/chrome-extension](https://github.com/clerk/javascript/blob/main/packages/chrome-extension/README.md) to work with SvelteKit, but it has resulted in no success. Additionally, it is worth noting that I have followed the steps outlined in the [@clerk/chrome-extension](https://github.com/clerk/javascript/blob/main/packages/chrome-extension/README.md) README, including disabling URL-based session sharing in dev/staging and adding the extension to `allowed_origins`.

I am not sure if I am settings my settings wrong or otherwise handling the session-sharing wrong. 

Please let me know if you have any advice, or if any more information would in evaluating the issue.

**Comment by aglontaru-sapiema** [comment]
Did you find any way of extending the "exp"? I also need to do this to allow my web app users to authenticate within a CLI.

**Reply by byte-my-briefs** [reply]
Not particularly. To fix my issue instead, I run the following function on an interval:

```
setTimeout(async () => {
        // Get a new, updated session cookie.
        const token = await clerk.session.getToken(...);

        // Update the cookie on my webapp for use by my extension.
        await chrome.cookies.set({
                httpOnly: false,
	        name: '__session',
	        path: '/',
	        secure: false,
	        url: import.meta.env.VITE_WEBAPP_URL,
	        value: token
        });
}, 30_000); // Run every 30s to ensure webapp cookie never goes stale
```

This exact approach will (of course) not work in a CLI because you do not have access to chrome API. However, perhaps there is a similar way you can use the Clerk client to fetch a new session token and manually update the cookie?

Happy to help brainstorm ideas and answer any questions, so please let me know!

**Reply by avemeva** [reply]
@ByteMyBriefs Hey! Can you please share a bit more on how to properly do that in Chrome Extension? For some reason my `clerk.session` is either `null`, or if I am using `clerk.load()` it complains about `no document` (seems like expected to be run in content script)

1. ~~Is this script part of content script or background script?~~ (background)
2. Can you show how you setup Clerk instance is that normal setup or with some things like in [chrome-extension](https://github.com/clerk/javascript/blob/9fa8831808fae5c97df18e78e7418bd08c4c8da9/packages/chrome-extension/src/singleton.ts#L21) package?

I have exactly the same situation as you do. I have Clerk on my website, but need to keep JWT refreshed in my Chrome Extension to access backend api.

**Reply by avemeva** [reply]
Link on [related discussion](https://discord.com/channels/856971667393609759/1256186396872413214) in Clerk discord

**Comment by jescalan** [comment]
https://github.com/orgs/clerk/discussions/3829

**Comment by thedogwiththedataonit** [comment]
I think I found the solution. The docs aren't too clear on this, but you have to create a JWT template to define a custom expiration. Once created, the getToken() function has a parameter called template which is where you put the name of the JWT template you created.

`
const token = await getToken({
                 template: "Admin",
})
`

I just spent 2 hours trying to figure this out so I hope it helps someone.

**Reply by jescalan** [reply]
Just want to add a couple notes here!

- If you need to extend the lifetime of the `__session` JWT, this is in fact the way to do it.
- It's worth noting however that this is an antipattern that we advise against doing, as this will result in a weaker security posture for your application.
- The session lifetime and expiration of the session JWT are two different things entirely. The session JWT expires every minute, but you aren't logged out when it expires, it is refreshed automatically in the background by Clerk.
- If you feel that you have a use case for your application that requires extending the session JWT expiration, I'd strongly recommend reaching out to Clerk support by emailing `support@clerk.com`, detailing your use case, and asking for a consult. In 99% of cases, there is a way you can make it work without compromising your app's security, and we can talk you through it.

**Reply by thedogwiththedataonit** [reply]
> Just want to add a couple notes here!
> 
> * If you need to extend the lifetime of the `__session` JWT, this is in fact the way to do it.
> * It's worth noting however that this is an antipattern that we advise against doing, as this will result in a weaker security posture for your application.
> * The session lifetime and expiration of the session JWT are two different things entirely. The session JWT expires every minute, but you aren't logged out when it expires, it is refreshed automatically in the background by Clerk.
> * If you feel that you have a use case for your application that requires extending the session JWT expiration, I'd strongly recommend reaching out to Clerk support by emailing `support@clerk.com`, detailing your use case, and asking for a consult. In 99% of cases, there is a way you can make it work without compromising your app's security, and we can talk you through it.

Hi Jeff! Thank you for your insight here. 

I am fairly new to Clerk so I am trying to wrap my head around how to best implement it. Would you suggest calling getToken() for every API call made from the client? I am currently storing it client side with the extended expiration, but if there is a better more secure approach here, I'd love to know! Thanks!

**Reply by jescalan** [reply]
Hey there! Appreciate the followup for sure. If your api is on the same origin (and subdomain) as your frontend, the token will be sent automatically with requests as a cookie. If you have set up your api on a separate domain or subdomain, all you need to do is make sure you are adding the token as a header to each request to your api. We have docs on how to do this [right here](https://clerk.com/docs/backend-requests/making/cross-origin#using-fetch-with-clerk-js), in case that's helpful!

**Reply by devransimsek** [reply]
Hi, Jeff!
What happens if we want to test our APIs on the Postman and we need user information? I think there should be a way to use for long time tokens at least on development.
@jescalan

**Reply by jescalan** [reply]
Generally, we recommend against trying to use longer time testing tokens, as JWTs cannot be revoked, so if you leak a long term token in any way, even if it's an accident, there's nothing you can do to stop whoever has the token from being able to fully take over that user's account short of rotating the instance keys and signing out all your users.

If you really feel strongly about this and are willing to accept the risks, you can create a custom JWT template and use this instead of the default session JWT in Clerk's dashboard. Otherwise, the best approach might be just to make sure that you [hit the `/tokens` endpoint](https://clerk.com/docs/reference/frontend-api/tag/Sessions#operation/createSessionToken) to get a fresh token before sending other test requests.

**Reply by shellscape** [reply]
@jescalan while that makes sense for production, it makes absolutely zero sense for the development tenant. there is no inherent risk to a development environment aside from a user treating it as production, and that is certainly not on Clerk. I'm a backend developer, and I don't dabble in the frontend. Using the API to constantly get refreshed tokens is unnecessarily burdensome for the use case, and Clerk is actively choosing to make DX worse for anyone not a frontend dev with that choice. If I want to say, test new endpoints using GraphiQL, I have to run a script constantly and typically before each request to the API, between server restarts, etc and I have to make sure that I copy the damn thing from output and paste it into an interface fast enough to execute a query before the thing expires.

I'm absolutely amazed that no one there has considered this scenario.

**Reply by jescalan** [reply]
Hey @shellscape - sorry for the super slow response here. I definitely understand what you're saying here. Maybe it would help for me to fill in some extra context here and get your take on what you'd recommend as a better flow as a backend developer. Entirely open to suggestions here!

Clerk is built a little differently from many other auth tools, in that it is, as you have noticed, a frontend-first tool. Interestingly, you don't actually need a backend at all to implement Clerk - it can be implemented on a fully client-rendered app with no server endpoints, and it works fine this way. If you do have a backend for your app, the intention is for requests from the front end to automatically pass along a JWT which can be verified without a network or database request, and that's it. The JWT architecture has its benefits - the speed gains from the previously mentioned "network-less" nature, as well as the ease of scaling due to it not being tied to databases, is great. However, it does also have downsides -- since it doesn't "phone home" at all, it's not possible to revoke a JWT. Clerk could declare a session revoked or expired in our database, but since you don't need to touch our database to auth your users, it wouldn't matter.

This inherently makes long-lived JWTs quite dangerous - if an attacker gets their hands on one, even if you realize this, it's too bad. There's nothing you can do about it until the JWT naturally expires. It's for this reason that Clerk uses very fast expiration JWTs - by naturally expiring the tokens quickly, not only does it effectively entirely burn the downside, in some situations, it makes it better. If an attacker gets their hands on a user's session token, it will automatically be dead in the water in an average of 30 seconds. With session tokens that are not JWTs and are verified via db call, you'd have to notice the attack and get the token revoked in less than that amount of time in order to be at an advantage, which I'd bet is quite unlikely in most cases. There's [a more in depth summary of how all this works here](https://clerk.com/docs/how-clerk-works/overview#main) in case its of interest.

In development, it's possible that these restrictions could be relaxed by default, but this would make the behavior of the system in development differ dramatically from production, which we have found is not desirable for our users. If you do want to go for it anyway, especially in cases where you're just testing endpoints, you could make a JWT template with a long expiration as suggested in the parent comment (we have made them with a 10 year expiration for this purpose sometimes) and use the JWT template instead of the default session token. You could then use the backend API to [generate a session token from the JWT template](https://clerk.com/docs/reference/backend-api/tag/Sessions#operation/CreateSessionTokenFromTemplate). This should give you back a token that would have a basically indefinite expiry that you could use for testing. I'd just caution anyone going in this direction to make sure you're only doing so for your development environment 😅

If anyone has feedback on what they think could be changed to make this a more pleasant experience without compromising security or prod/dev environment similarity, please feel free to drop a comment and I'll do what I can to get it implemented if its a good idea! Regardless, I am gonna make sure we have a page on how to do this in the documentation so it's easier to find.

**Reply by shellscape** [reply]
Appreciate the writeup. Allowing users in development to enter a danger zone setting and providing longer-lived tokens only in that environment still seems like that the right path. Especially so since you have very clear guardrails around your development environments (which is a great call btw). The backend developer definitely needs more consideration here, and providing models for that developer is important.

Barring some sort of danger zone setting, if the folks there just aren't comfortable with allowing users to make a potentially bad decision and want to provide some additional protection, there's an alternate path: provide a default long-lived template that's only available in dev, and a component that people can plop onto their site which exposes the result of getting that token from the long-lived template. That's essentially what we've done in the meantime, and it's on a page of the frontend app which is only accessible in dev. That let's me, as a backend developer, continually test our backend services with minimal burden. Given that this is tangential to the recommended path, and touches on what you've written above, I don't see why Clerk wouldn't wire this up to provide a component to people ready to go for that use case. There seems to be enough discussion around this on the webs to warrant it. At the very least, some clear documentation in the official docs to address the use case.

**Reply by jescalan** [reply]
Super helpful context here, really appreciate it. I'll make sure that this is discussed internally and hopefully we can come up with something good to make this situation better for devs!

**Reply by jescalan** [reply]
@shellscape - want to check in here quickly - would [this backed api endpoint](https://clerk.com/docs/reference/backend-api/tag/Sessions#operation/CreateSessionToken) do the trick for you? it allows creation of a session token from the backend for testing purposes, and you can set a custom expiration.

**Reply by shellscape** [reply]
outstanding, thank you.

**Reply by jescalan** [reply]
Hooray! I will make sure this is documented more prominently. Let me ask here since you might have a good take - if you were going into our docs to look for something like this, what would you search for?

**Reply by T-P-F** [reply]
@jescalan I've been searching for exactly this so here are some searches I did (including the one that led me to this thread): "session-token-expired-refresh-non-eligible-non-get" "x-clerk-auth-reason" "clerk session expiration" "clerk token expiration" "clerk dev environment tokens" etc..

**Reply by jescalan** [reply]
Please [reach out to Clerk support](https://clerk.com/contact/support) if you are having issues! We do not monitor or reply to github discussions.
